Add a Prisma tutorial (#131)
dbos-inc/dbos-docs: docs/tutorials/using-prisma.md

DBOS supports Prisma as an alternative to [Knex](https://knexjs.org/) and [TypeORM](https://typeorm.io) for transactional data management.

### Getting Started

An easy way to get started with Prisma is to bootstrap your application with our Prisma template.
This is similar to the template used in the [quickstart](../getting-started/quickstart.md), but built with Prisma instead of Knex.
To download it, run:

```bash
npx -y @dbos-inc/create@latest -t hello-prisma -n <app-name>
```

Then, build it, run schema migrations, and start the Prisma sample app:

```bash
npm run build
npx dbos migrate
npx dbos start
```

To see that it's working, visit this URL in your browser: [http://localhost:3000/greeting/dbos](http://localhost:3000/greeting/dbos).  You should get this message: `Greeting 1: Hello, dbos!` Each time you refresh the page, the counter should go up by one.

### Setting Up Prisma Schema

In DBOS, the Prisma Schema is defined in [the same way as any other Prisma project](https://www.prisma.io/docs/orm/prisma-schema/overview).
In this tutorial we assume it is located at the default path (`prisma/schema.prisma`), for example:

> 

Reviewer: Do we not support ```prisma ?

> ```javascript title="prisma/schema.prisma"

Reviewer: Do we not support ```prisma ?

Author: No syntax highlighting for Prisma so far (we use Docusaurus built-in list of supported languages)

```javascript title="generate_env.js"
const { parseConfigFile } = require('@dbos-inc/dbos-sdk/dist/src/dbos-runtime/config');
const fs = require('node:fs');
const path = require('node:path');

// Load the configuration file
const [dbosConfig, ] = parseConfigFile();

// Write out the .env file
const databaseURL = `postgresql://${dbosConfig.poolConfig.user}:${dbosConfig.poolConfig.password}@${dbosConfig.poolConfig.host}:${dbosConfig.poolConfig.port}/${dbosConfig.poolConfig.database}`;

try {
  fs.writeFileSync(path.join(process.cwd(), 'prisma', '.env'), `DATABASE_URL="${databaseURL}"`);
  console.log("Wrote database URL to the prisma/.env file.");
} catch (error) {
  console.error("Error writing prisma/.env file:", error.message);
}
```

You also need to generate a [Prisma Client](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/introduction) from your Prisma schema before compiling your application.
Therefore, we recommend that you set the build script in your `package.json` file as follows:
```js title="package.json"

> {

Reviewer: ```json ??

Author: ```json doesn't have syntax highlighting.

---
sidebar_position: 12
title: Using Prisma
description: Learn how to use Prisma schema and perform transactional updates
---

## Prisma Overview
[Prisma](https://www.prisma.io/) is a popular open-source TypeScript ORM.
The main idea is to define your application data models (entities) in [Prisma Schema](https://www.prisma.io/docs/orm/prisma-schema/overview), and then use Prisma to automatically generate [schema migrations](https://www.prisma.io/docs/orm/prisma-migrate/getting-started) as well as the corresponding [Prisma Client](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/introduction) for you to manage your data.
DBOS supports Prisma as an alternative to [Knex](https://knexjs.org/) and [TypeORM](https://typeorm.io) for transactional data management.

## Usage

> DBOS supports essentially direct use of Prisma, but a few additional steps are necessary to generate schema migrations and the client, and use a transaction in the context of a workflow.

Reviewer 2: ```suggestion
DBOS supports direct use of Prisma with only a few additional configuration steps to generate schema migrations and the Prisma client. The client can then be used in the context of a workflow.
```

### Getting Started

An easy way to get started with Prisma is to bootstrap your application with our Prisma template.
This is similar to the template used in the [quickstart](../getting-started/quickstart.md), but built with Prisma instead of Knex.
To download it, run:

```bash
npx -y @dbos-inc/create@latest -t hello-prisma -n <app-name>
```

Then, build it, run schema migrations, and start the Prisma sample app:

```bash
npm run build
npx dbos migrate
npx dbos start
```

To see that it's working, visit this URL in your browser: [http://localhost:3000/greeting/dbos](http://localhost:3000/greeting/dbos).  You should get this message: `Greeting 1: Hello, dbos!` Each time you refresh the page, the counter should go up by one.

### Setting Up Prisma Schema

> In DBOS, Prisma Schema by default is located at `./prisma/schema.prisma` and is defined in [the same way as any other Prisma project](https://www.prisma.io/docs/orm/prisma-schema/overview), for example:

Reviewer 2: ```suggestion
In DBOS, Prisma Schema is located, by default, at `./prisma/schema.prisma` and is defined in [the same way as any other Prisma project](https://www.prisma.io/docs/orm/prisma-schema/overview), for example:
```

```javascript
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model DbosHello {
  @@map("dbos_hello")
  greeting_id Int @id @default(autoincrement())
  greeting String
}
```

Note that the schema file requires a `$DATABASE_URL` environment variable to connect to a database. To make it easy, we include in our template a script to automatically generate a `.env` file under the `./prisma` folder with the correct database URL constructed from your `dbos-config.yaml` file, please see the [document below](#configuring-prisma).

### Schema Management

In production scenarios or when using DBOS Cloud, we strongly recommend you manage your database schema using migrations.

> [Prisma Migrate](https://www.prisma.io/docs/orm/prisma-migrate/getting-started) provides plenty of support for generating and managing migrations.

Reviewer 2: ```suggestion
Make sure to check [Prisma Migrate](https://www.prisma.io/docs/orm/prisma-migrate/getting-started) documentation for generating and managing Prisma migrations.
```

export class Hello {
  @Transaction()
  static async helloTransaction(txnCtxt: TransactionContext<PrismaClient>, name: string)  {
    const greeting = `Hello, ${name}!`;
    const res = await txnCtxt.client.dbosHello.create({
      data: {
        greeting: greeting,
      },
    });
    return `Greeting ${res.greeting_id}: ${greeting}`;
  }
}
```

If preferred, it is possible to define a `type` to clean up the transaction method prototypes a little bit.
```javascript
type PrismaTransactionContext = TransactionContext<PrismaClient>;
```

### Configuring Prisma

:::info
If you are using the [Prisma template](#getting-started), this is done for you.
:::

To enable Prisma, you must set the `app_db_client` field in the [DBOS configuration file](../api-reference/configuration.md) to `prisma`.
You should also configure Prisma migration commands.
Here is an example of a configuration file set up for Prisma:

```yaml
database:
  hostname: 'localhost'
  port: 5432
  username: 'postgres'
  app_db_name: 'hello_prisma'
  password: ${PGPASSWORD}
  connectionTimeoutMillis: 3000
  app_db_client: prisma
  migrate:
    - npx prisma migrate deploy
```

:::info

> Prisma doesn't [support migration rollback](https://github.com/prisma/prisma/discussions/4617) for successfully applied migrations like other ORMs.

Reviewer 2: ```suggestion
Prisma doesn't [support schema rollbacks](https://github.com/prisma/prisma/discussions/4617) for successfully applied migrations like other ORMs.
```

Author: It's schema migration rollback.
